File: cpp/core/Metrics.cpp
// See <https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/>
// for docs.
#include "Metrics.hpp"

#include "Assert.hpp"
#define CPPHTTPLIB_USE_POLL
#include "httplib.h"

// We're being overly conservative here, see
// <https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/#special-characters>,
// but better being safe than sorry.
static void validMetricsName(const std::string& name) {
    for (auto ch : name) {
        ALWAYS_ASSERT(
            (ch >= 'a' && ch <= 'z') ||
            (ch >= 'A' && ch <= 'Z') ||
            (ch >= '0' && ch <= '9') ||
            ch == '-' || ch == '_' || ch == '.' || ch == ':'
        );
    }
}

void MetricsBuilder::measurement(const std::string& name) {
    ALWAYS_ASSERT(_state == State::INIT || _state == State::TIMESTAMP);
    validMetricsName(name);
    if (_state == State::TIMESTAMP) {
        _payload += '\n';
    }
    _payload += name;
    _state = State::MEASUREMENT;
}

void MetricsBuilder::tagRaw(const std::string& name, const std::string& value) {
    ALWAYS_ASSERT(_state == State::MEASUREMENT || _state == State::TAGS);
    validMetricsName(name);
    validMetricsName(value);
    _payload += ',';
    _payload += name;
    _payload += '=';
    _payload += value;
    _state = State::TAGS;
}

void MetricsBuilder::fieldRaw(const std::string& name, const std::string& value) {
    ALWAYS_ASSERT(_state == State::MEASUREMENT || _state == State::TAGS || _state == State::FIELDS);
    validMetricsName(name);
    if (_state == State::MEASUREMENT || _state == State::TAGS) {
        _payload += ' ';
    } else {
        _payload += ',';
    }
    _payload += name;
    _payload += '=';
    _payload += value;
    _state = State::FIELDS;
}

void MetricsBuilder::timestamp(TernTime t) {
    ALWAYS_ASSERT(_state == State::FIELDS);
    _payload += ' ';
    static char buf[21];
    std::sprintf(buf, "%lu", t.ns);
    _payload += buf;
    _state = State::TIMESTAMP;
}

std::string sendMetrics(const InfluxDB& idb, Duration timeout, const std::string& payload) {
    httplib::Client cli(idb.origin);
    time_t ds = timeout.ns / 1'000'000'000ull;
    time_t dus = (timeout.ns % 1'000'000'000ull) / 1'000ull;
    cli.set_connection_timeout(ds, dus);
    cli.set_read_timeout(ds, dus);
    cli.set_write_timeout(ds, dus);

    std::string path = "/api/v2/write?org=" + idb.org + "&bucket=" + idb.bucket + "&precision=ns";
    const auto res = cli.Post(path, payload, "text/plain");
    if (res.error() != httplib::Error::Success) {
        std::ostringstream ss;
        ss << "Could not insert metrics to " << idb.origin << path << ": " << res.error();
        return ss.str();
    }
    return {};
}

File: go/core/log/metrics.go
// Copyright 2025 XTX Markets Technologies Limited
//
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

// See <https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/>
// for docs.
package log

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"time"
)

const (
	METRICS_INIT        uint8 = 0
	METRICS_MEASUREMENT uint8 = 1
	METRICS_TAGS        uint8 = 2
	METRICS_FIELDS      uint8 = 3
	METRICS_TIMESTAMP   uint8 = 4
)

type MetricsBuilder struct {
	state   uint8
	payload bytes.Buffer
}

func (b *MetricsBuilder) Reset() {
	b.state = METRICS_INIT
	b.payload.Reset()
}

func (b *MetricsBuilder) Payload() io.Reader {
	if !(b.state == METRICS_INIT || b.state == METRICS_TIMESTAMP) {
		panic(fmt.Errorf("bad state %v for payload", b.state))
	}
	return bytes.NewReader(b.payload.Bytes())
}

// We're being overly conservative here, see
// <https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/#special-characters>,
// but better being safe than sorry.
func validMetricsName(name string) {
	for _, ch := range name {
		if !((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '-' || ch == '_' || ch == '.' || ch == ':') {
			panic(fmt.Errorf("bad metrics name %q", name))
		}
	}
}

func (b *MetricsBuilder) Measurement(name string) {
	if !(b.state == METRICS_INIT || b.state == METRICS_TIMESTAMP) {
		panic(fmt.Errorf("bad state %v for measurement", b.state))
	}
	validMetricsName(name)
	if b.state == METRICS_TIMESTAMP {
		b.payload.Write([]byte{'\n'})
	}
	b.payload.Write([]byte(name))
	b.state = METRICS_MEASUREMENT
}

func (b *MetricsBuilder) Tag(name string, value string) {
	if !(b.state == METRICS_MEASUREMENT || b.state == METRICS_TAGS) {
		panic(fmt.Errorf("bad state %v for tag", b.state))
	}
	validMetricsName(name)
	validMetricsName(value)
	fmt.Fprintf(&b.payload, ",%s=%s", name, value)
	b.state = METRICS_TAGS
}

func (b *MetricsBuilder) fieldRaw(name string, value string) {
	if !(b.state == METRICS_MEASUREMENT || b.state == METRICS_TAGS || b.state == METRICS_FIELDS) {
		panic(fmt.Errorf("bad state %v for field", b.state))
	}
	validMetricsName(name)
	if b.state == METRICS_MEASUREMENT || b.state == METRICS_TAGS {
		b.payload.Write([]byte{' '})
	} else {
		b.payload.Write([]byte{','})
	}
	fmt.Fprintf(&b.payload, "%s=%s", name, value)
	b.state = METRICS_FIELDS

}

func (b *MetricsBuilder) FieldU64(name string, value uint64) {
	b.fieldRaw(name, fmt.Sprintf("%vi", value))
}

func (b *MetricsBuilder) FieldU32(name string, value uint32) {
	b.fieldRaw(name, fmt.Sprintf("%vi", value))
}

func (b *MetricsBuilder) Timestamp(t time.Time) {
	if !(b.state == METRICS_FIELDS) {
		panic(fmt.Errorf("bad state %v for timestamp", b.state))
	}
	fmt.Fprintf(&b.payload, " %v", t.UnixNano())
	b.state = METRICS_TIMESTAMP
}

type InfluxDB struct {
	Origin string
	Org    string
	Bucket string
}

func (idb *InfluxDB) SendMetrics(payload io.Reader) error {
	url := idb.Origin + "/api/v2/write?org=" + idb.Org + "&bucket=" + idb.Bucket + "&precision=ns"
	resp, err := http.Post(url, "text/plain", payload)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return nil
}
